File: ios-app/Targets/BombestBeats/Sources/Views/Components/OscilloscopeVisualizer.swift
import SwiftUI

/// Oscilloscope-style waveform visualizer for Studio Dust theme
/// Features thin glowing lines with natural imperfections, phosphor glow effect
struct OscilloscopeVisualizer: View {
    let amplitudes: [Float]
    var primaryColor: Color = Color(hex: 0xD4A574)
    var glowColor: Color = Color(hex: 0x5A7D7E)
    var lineThickness: CGFloat = 2
    
    var body: some View {
        TimelineView(.animation(minimumInterval: 1/30)) { timeline in
            Canvas { context, size in
                let width = size.width
                let height = size.height
                let centerY = height / 2
                
                guard !amplitudes.isEmpty else { return }
                
                let pointCount = amplitudes.count
                let stepX = width / CGFloat(pointCount - 1).clamped(to: 1...CGFloat.greatestFiniteMagnitude)
                
                // Calculate energy for dynamic effects
                let energy = CGFloat(amplitudes.reduce(0, +) / Float(amplitudes.count))
                
                // Build points with natural jitter
                var points: [CGPoint] = []
                let time = Date().timeIntervalSince1970
                
                for (i, amp) in amplitudes.enumerated() {
                    let x = CGFloat(i) * stepX
                    let clampedAmp = CGFloat(min(max(amp, 0), 1))
                    
                    // Add natural jitter
                    let jitter = CGFloat.random(in: -2...2)
                    let y = centerY - (clampedAmp * height * 0.4) + jitter
                    
                    points.append(CGPoint(x: x, y: y))
                }
                
                guard points.count >= 2 else { return }
                
                // Draw glow layers
                for glowRadius in [8.0, 5.0, 3.0] as [CGFloat] {
                    let glowPath = createSmoothPath(points: points)
                    let glowAlpha = (0.1 / (glowRadius / 3)) * (0.5 + energy * 0.5)
                    
                    context.stroke(
                        glowPath,
                        with: .color(glowColor.opacity(Double(glowAlpha.clamped(to: 0...0.3)))),
                        style: StrokeStyle(
                            lineWidth: glowRadius + lineThickness,
                            lineCap: .round,
                            lineJoin: .round
                        )
                    )
                }
                
                // Main waveform line
                let mainPath = createSmoothPath(points: points)
                let dynamicThickness = lineThickness * (0.8 + energy * 0.4)
                
                context.stroke(
                    mainPath,
                    with: .color(primaryColor.opacity(0.9)),
                    style: StrokeStyle(
                        lineWidth: dynamicThickness,
                        lineCap: .round,
                        lineJoin: .round
                    )
                )
                
                // Phosphor bright center
                context.stroke(
                    mainPath,
                    with: .color(Color.white.opacity(Double(0.4 + energy * 0.3))),
                    style: StrokeStyle(
                        lineWidth: dynamicThickness * 0.3,
                        lineCap: .round,
                        lineJoin: .round
                    )
                )
                
                // Draw scan lines for CRT effect
                drawScanLines(context: context, width: width, height: height, opacity: 0.03)
                
                // Draw grain overlay
                drawGrain(context: context, width: width, height: height, opacity: 0.02)
            }
        }
    }
    
    /// Create smooth bezier path through points
    private func createSmoothPath(points: [CGPoint]) -> Path {
        var path = Path()
        path.move(to: points[0])
        
        for i in 1..<points.count {
            let prev = points[i - 1]
            let curr = points[i]
            let controlX = (prev.x + curr.x) / 2
            let controlY = (prev.y + curr.y) / 2
            
            path.addQuadCurve(to: CGPoint(x: controlX, y: controlY), control: prev)
        }
        
        if let last = points.last {
            path.addLine(to: last)
        }
        
        return path
    }
    
    /// Draw horizontal scan lines
    private func drawScanLines(context: GraphicsContext, width: CGFloat, height: CGFloat, opacity: Double) {
        let spacing: CGFloat = 4
        var y: CGFloat = 0
        
        while y < height {
            var path = Path()
            path.move(to: CGPoint(x: 0, y: y))
            path.addLine(to: CGPoint(x: width, y: y))
            
            context.stroke(
                path,
                with: .color(Color.black.opacity(opacity)),
                lineWidth: 1
            )
            y += spacing
        }
    }
    
    /// Draw subtle grain overlay
    private func drawGrain(context: GraphicsContext, width: CGFloat, height: CGFloat, opacity: Double) {
        for _ in 0..<30 {
            let x = CGFloat.random(in: 0...width)
            let y = CGFloat.random(in: 0...height)
            let size = CGFloat.random(in: 0.5...2)
            
            context.fill(
                Path(ellipseIn: CGRect(x: x, y: y, width: size, height: size)),
                with: .color(Color.white.opacity(opacity * Double.random(in: 0...1)))
            )
        }
    }
}

// MARK: - Comparable Extension for Clamping

extension Comparable {
    func clamped(to range: ClosedRange<Self>) -> Self {
        return min(max(self, range.lowerBound), range.upperBound)
    }
}

File: android-app/app/src/main/java/com/bombest/music/ui/components/OscilloscopeVisualizer.kt
package com.bombest.music.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.*
import androidx.compose.ui.graphics.drawscope.*
import kotlin.math.*
import kotlin.random.Random

/**
 * Oscilloscope-style waveform visualizer.
 * Features:
 * - Thin glowing lines with natural imperfections
 * - Dynamic reaction to audio energy
 * - Slight jitter and uneven thickness for analog feel
 * - No symmetry enforcement - organic movement
 * - Phosphor glow effect like vintage CRT
 */
@Composable
fun OscilloscopeVisualizer(
    amplitudes: FloatArray,
    modifier: Modifier = Modifier,
    primaryColor: Color = Color(0xFFD4A574),
    glowColor: Color = Color(0xFF5A7D7E),
    lineThickness: Float = 2f
) {
    // Calculate average energy for intensity
    val energy = remember(amplitudes) {
        if (amplitudes.isEmpty()) 0f else amplitudes.average().toFloat()
    }
    
    Canvas(modifier = modifier.fillMaxSize()) {
        val width = size.width
        val height = size.height
        val centerY = height / 2
        
        if (amplitudes.isEmpty()) return@Canvas
        
        val pointCount = amplitudes.size
        val stepX = width / (pointCount - 1).coerceAtLeast(1)
        
        // Create path for the waveform
        val path = Path()
        val glowPath = Path()
        
        // Random seed for consistent jitter per frame
        val jitterSeed = System.currentTimeMillis() / 50 // Changes every 50ms
        val random = Random(jitterSeed.toInt())
        
        // Build waveform points with natural imperfections
        val points = mutableListOf<Offset>()
        
        for (i in amplitudes.indices) {
            val x = i * stepX
            
            // Base amplitude
            val amp = amplitudes[i].coerceIn(0f, 1f)
            
            // Add natural jitter (minor imperfections)
            val jitter = (random.nextFloat() - 0.5f) * 4f
            
            // Scale to canvas with organic variance
            val y = centerY - (amp * height * 0.4f) + jitter
            
            points.add(Offset(x, y))
        }
        
        // Draw waveform using smooth curves
        if (points.size >= 2) {
            // Outer glow layer (blurred effect simulated with multiple strokes)
            for (glowRadius in listOf(8f, 5f, 3f)) {
                path.reset()
                path.moveTo(points.first().x, points.first().y)
                
                for (i in 1 until points.size) {
                    val prevPoint = points[i - 1]
                    val currPoint = points[i]
                    
                    // Smooth curve between points
                    val controlX = (prevPoint.x + currPoint.x) / 2
                    path.quadraticBezierTo(
                        prevPoint.x, prevPoint.y + (random.nextFloat() - 0.5f) * 2,
                        controlX, (prevPoint.y + currPoint.y) / 2
                    )
                }
                
                // Complete to last point
                path.lineTo(points.last().x, points.last().y)
                
                val glowAlpha = (0.1f / (glowRadius / 3f)) * (0.5f + energy * 0.5f)
                drawPath(
                    path = path,
                    color = glowColor.copy(alpha = glowAlpha.coerceIn(0f, 0.3f)),
                    style = Stroke(
                        width = glowRadius + lineThickness,
                        cap = StrokeCap.Round,
                        join = StrokeJoin.Round
                    )
                )
            }
            
            // Main waveform line
            path.reset()
            path.moveTo(points.first().x, points.first().y)
            
            for (i in 1 until points.size) {
                val prevPoint = points[i - 1]
                val currPoint = points[i]
                
                val controlX = (prevPoint.x + currPoint.x) / 2
                path.quadraticBezierTo(
                    prevPoint.x, prevPoint.y,
                    controlX, (prevPoint.y + currPoint.y) / 2
                )
            }
            path.lineTo(points.last().x, points.last().y)
            
            // Dynamic thickness based on energy
            val dynamicThickness = lineThickness * (0.8f + energy * 0.4f)
            
            drawPath(
                path = path,
                color = primaryColor.copy(alpha = 0.9f),
                style = Stroke(
                    width = dynamicThickness,
                    cap = StrokeCap.Round,
                    join = StrokeJoin.Round
                )
            )
            
            // Bright center line for phosphor effect
            drawPath(
                path = path,
                color = Color.White.copy(alpha = 0.4f + energy * 0.3f),
                style = Stroke(
                    width = dynamicThickness * 0.3f,
                    cap = StrokeCap.Round,
                    join = StrokeJoin.Round
                )
            )
        }
        
        // Add subtle scan lines for CRT effect
        drawScanLines(height, width, 0.03f)
        
        // Add subtle grain
        drawGrainOverlay(size.width, size.height, 0.02f)
    }
}

/**
 * Draw horizontal scan lines for CRT effect
 */
private fun DrawScope.drawScanLines(height: Float, width: Float, opacity: Float) {
    val lineSpacing = 4f
    var y = 0f
    
    while (y < height) {
        drawLine(
            color = Color.Black.copy(alpha = opacity),
            start = Offset(0f, y),
            end = Offset(width, y),
            strokeWidth = 1f
        )
        y += lineSpacing
    }
}

/**
 * Draw subtle grain overlay
 */
private fun DrawScope.drawGrainOverlay(width: Float, height: Float, opacity: Float) {
    val random = Random(System.currentTimeMillis() / 100)
    val particleCount = 30
    
    for (i in 0 until particleCount) {
        val x = random.nextFloat() * width
        val y = random.nextFloat() * height
        val size = random.nextFloat() * 1.5f + 0.5f
        
        drawCircle(
            color = Color.White.copy(alpha = opacity * random.nextFloat()),
            radius = size,
            center = Offset(x, y)
        )
    }
}
